unit-tests/utilities/concurrency/test-usage.cpp

```diff
@@ -0,0 +1,194 @@
+    stopwatch sw;
+
+    sw.reset();
+
+    // dequeue an item after 5 seconds, we wait for the blocking call and verify we return after this dequeue call ~ 5 seconds
+    std::thread dequeue_thread( [&]() {
+        std::this_thread::sleep_for( std::chrono::seconds( 5 ) );
+        scq.dequeue( &f, 1000 );
+    } );
+
+    REQUIRE( sw.get_elapsed_ms() < 1000 );
+
+    for( int i = 0; i < 10; ++i )
+    {
+        scq.blocking_enqueue( []() {} );
+    }
+    REQUIRE( sw.get_elapsed_ms() < 1000 );
+    REQUIRE( scq.size() == 10 );            // verify queue is full (default capacity is 10)
+    scq.blocking_enqueue( []() {} );        // add a blocking call (item 11)
+    REQUIRE( sw.get_elapsed_ms() > 5000 );  // verify the blocking call return on dequeue time
+    REQUIRE( sw.get_elapsed_ms() < 6000 );  // verify the blocking call return on dequeue time
+
+    dequeue_thread.join();  // verify clean exit with no threads alive.
+}
+
+TEST_CASE("verify mutex protection")
+{
+    single_consumer_queue< int > scq;
+    stopwatch sw;
+
+    const int MAX_SIZE_FOR_THREAD = 20;
+    std::thread enqueue_thread1( [&]() {
+        for( int i = 0; i < MAX_SIZE_FOR_THREAD; ++i )
+        {
+            std::this_thread::sleep_for( std::chrono::milliseconds( 100 ) );
+            scq.blocking_enqueue(std::move(i));
+        }
+    } );
+
+    std::thread enqueue_thread2( [&]() {
+        for( int i = 0; i < MAX_SIZE_FOR_THREAD; ++i )
+        {
+            std::this_thread::sleep_for( std::chrono::milliseconds( 100 ) );
+            scq.blocking_enqueue(std::move(20 + i));
+        }
+    } );
+
+    std::vector<int> all_values;
+    for( int i = 0; i < MAX_SIZE_FOR_THREAD * 2; ++i )
+    {
+        int val;
+        scq.dequeue( &val, 1000 );
+        all_values.push_back(val);
+    }
+
+    REQUIRE(all_values.size() == MAX_SIZE_FOR_THREAD * 2);
+
+    std::sort(all_values.begin(), all_values.end());
+
+    for (int i = 0; i < all_values.size(); ++i)
+    {
+        REQUIRE(all_values[i] == i);
+    }
+
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+
+    enqueue_thread1.join();
+    enqueue_thread2.join();
+
+}
+
+TEST_CASE("verify stop() not consuming high CPU usage")
+{
+    // using shared_ptr because no copy constructor is allowed for a dispatcher.
+    std::cout << "starting dispatchers()\n";
+    std::vector<std::shared_ptr<dispatcher>> dispatchers;
+    stopwatch sw;
+
+
+    for (int i = 0 ; i < 32; ++i)
+    {
+        dispatchers.push_back(std::make_shared<dispatcher>(10));
+    }
+
+    for (auto &&dispatcher : dispatchers)
+    {
+        dispatcher->start();
+    }
+
+    for (auto&& dispatcher : dispatchers)
+    {
+        dispatcher->stop();
+    }
+
+  
+    // Allow some time for all threads to do some work
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+
+    sw.reset();
+
+    // Do some stress work
+    REQUIRE(fibo(40) == 165580141);
+    // Verify the stress test did not take too long.
```

Comment:
I think we need to describe what we saw when this test did not pass...

Reply:
Added